yarn-project/simulator/src/public/benchmarker.ts

```diff
@@ -0,0 +1,54 @@
+import { type L2BlockBuiltStats } from '@aztec/circuit-types/stats';
+import { elapsed } from '@aztec/foundation/timer';
+import { type ToCamelCase } from '@aztec/foundation/types';
+
+import { type PublicKernelPhase } from './abstract_phase_manager.js';
+
+type Event = `kernel-simulation` | `avm-simulation`;
+type PhaseEvent = `${PublicKernelPhase}-${Event}`;
+type PublicEventDurations = Record<PhaseEvent, number[]>;
+
+// exported benchmarks need to be in camel case
+type PhaseEventBenchmark = `${ToCamelCase<PhaseEvent>}AvgMs`;
+
+function camelCasePhaseEvent(phaseEvent: PhaseEvent): PhaseEventBenchmark {
+  const camelCased = phaseEvent.replace(/-([a-z])/g, g => g[1].toUpperCase());
+  return `${camelCased}AvgMs` as PhaseEventBenchmark;
+}
+
+type AverageDurations = Record<PhaseEventBenchmark, number>;
+
+export class PublicBenchmarker {
+  public durations: PublicEventDurations;
+
+  constructor() {
+    this.durations = {
+      'setup-kernel-simulation': [],
+      'setup-avm-simulation': [],
+      'app-logic-kernel-simulation': [],
+      'app-logic-avm-simulation': [],
+      'teardown-kernel-simulation': [],
+      'teardown-avm-simulation': [],
+      'tail-kernel-simulation': [],
+      'tail-avm-simulation': [],
+    };
+  }
+
+  async benchmark<T>(phase: PublicKernelPhase, event: Event, callback: () => T) {
+    const [durationMs, result] = await elapsed(callback);
+    this.durations[`${phase}-${event}`].push(durationMs);
+    return result;
+  }
+
+  getAverageDurations(): AverageDurations {
```

Comment:
The stats module already averages all the values for a metric so I think this leads to an average of averages (which most likely isn't what we want?).

https://github.com/AztecProtocol/aztec-packages/blob/39747b933a13aa08f25c5074207f9d92489d5e3d/yarn-project/scripts/src/benchmarks/aggregate.ts#L295-L305